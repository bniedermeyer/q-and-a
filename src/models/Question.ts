export interface Question {
  /**
   * The question the user asked
   */
  question: string;
  /**
   * Optional user identifier
   */
  userId?: string;
  /**
   * Optional correlation id
   */
  correlationId?: string;
}

export interface AskedQuestion extends Question {
  /**
   * Count of times the question has been asked. Must be present on all questions retrieved from the server
   */
  count?: number;
  /**
   * Unique question identifier. Must be present on all questions retrieved from the server.
   */
  key?: string;
}
