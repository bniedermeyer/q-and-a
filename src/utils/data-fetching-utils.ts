import { AskedQuestion, Question } from '../models/Question';
import state from '../store';

interface IncrementRequest {
  key: string;
  userId?: string;
}

/**
 * Posts the question asked by the user to the backend.
 *
 * @param question The question asked by the user
 */
export async function askQuestion(question: string): Promise<void> {
  let request: Question = {
    question,
  };
  if (state.userId) {
    request.userId = state.userId;
  }
  if (state.correlationId) {
    request.correlationId = state.correlationId;
  }
  const settings = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  };
  await fetch(state.askEndpoint, settings);
}

/**
 * Posts a command to the backend to increment the number of times the question has been asked.
 *
 * @param key The unique id key of the question to increment
 */
export async function incrementQuestionCount(key: string): Promise<void> {
  const request: IncrementRequest = {
    key,
  };
  if (state.userId) {
    request.userId = state.userId;
  }

  const settings = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  };
  await fetch(state.askEndpoint, settings);
}

/**
 * Requests all questions from the backend and updates the state with the result.
 * If the component is initialized with a correlation ID, this will pass the id in the request to only retrieve questions what are related to that ID.
 */
export async function fetchQuestions(): Promise<void> {
  const url = state.correlationId ? `${state.retrieveEndpoint}?correlationId=${state.correlationId}` : state.retrieveEndpoint;
  let data = await fetch(url);
  let questions: AskedQuestion[] = await data.json();
  state.questions = questions;
}
