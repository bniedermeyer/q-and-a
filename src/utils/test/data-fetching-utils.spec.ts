import fetchMock from 'fetch-mock-jest';
import { askQuestion, incrementQuestionCount, fetchQuestions } from '../data-fetching-utils';

import state, { disposeStore } from '../../store';
import { AskedQuestion } from '../../models/Question';

beforeEach(() => {
  fetchMock.mockReset();
  disposeStore();
});

describe('askQuestion', () => {
  beforeEach(() => {
    fetchMock.post(state.askEndpoint, (_url, options) => {
      if (options.body.length > 0) {
        return 200;
      }
      return 400;
    });
  });

  it('should post the question to the backend', async () => {
    await askQuestion('What is the airspeed velocity of an unladen swallow?');
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('should post the question to the backend with the expected metadata', async () => {
    state.userId = 'my-user-id';
    state.correlationId = 'my-correlation-id';
    await askQuestion('What is the airspeed velocity of an unladen swallow?');
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });
});

describe('incrementQuestionCount', () => {
  beforeEach(() => {
    fetchMock.post(state.askEndpoint, (_url, options) => {
      if (options.body.length > 0) {
        return 200;
      }
      return 400;
    });
  });

  it('should post the increment request to the backend', async () => {
    await incrementQuestionCount('question-key');
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('should post the increment request to the backend with the expected metadata', async () => {
    state.userId = 'my-user-id';
    await incrementQuestionCount('question-key');
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });
});

describe('fetchQuestions', () => {
  const questions: AskedQuestion[] = [
    { question: 'What time is it?', count: 2, key: '1', userId: 'user1', correlationId: 'a123' },
    { question: 'What is for lunch?', count: 0, key: '2', userId: '1', correlationId: 'a123' },
  ];
  it('should fetch the questions and set them in the state', async () => {
    fetchMock.get(
      state.retrieveEndpoint,
      questions.map(rawQuestion => {
        const { correlationId, ...question } = rawQuestion;
        return question;
      }),
    );
    await fetchQuestions();
    expect(state.questions).toMatchSnapshot();
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });

  it('should fetch the questions for the given correlation id and set them in the state', async () => {
    state.correlationId = 'a123';
    fetchMock.get(`${state.retrieveEndpoint}?correlationId=${state.correlationId}`, questions);
    await fetchQuestions();
    expect(state.questions).toMatchSnapshot();
    expect(fetchMock.mock.calls).toMatchSnapshot();
  });
});
