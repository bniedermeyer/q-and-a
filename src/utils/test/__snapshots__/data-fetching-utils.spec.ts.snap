// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`askQuestion should post the question to the backend 1`] = `
Array [
  Array [
    "/ask",
    Object {
      "body": "{\\"question\\":\\"What is the airspeed velocity of an unladen swallow?\\"}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`askQuestion should post the question to the backend with the expected metadata 1`] = `
Array [
  Array [
    "/ask",
    Object {
      "body": "{\\"question\\":\\"What is the airspeed velocity of an unladen swallow?\\",\\"userId\\":\\"my-user-id\\",\\"correlationId\\":\\"my-correlation-id\\"}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchQuestions should fetch the questions and set them in the state 1`] = `
Array [
  Object {
    "count": 2,
    "key": "1",
    "question": "What time is it?",
    "userId": "user1",
  },
  Object {
    "count": 0,
    "key": "2",
    "question": "What is for lunch?",
    "userId": "1",
  },
]
`;

exports[`fetchQuestions should fetch the questions and set them in the state 2`] = `
Array [
  Array [
    "/questions",
  ],
]
`;

exports[`fetchQuestions should fetch the questions for the given correlation id and set them in the state 1`] = `
Array [
  Object {
    "correlationId": "a123",
    "count": 2,
    "key": "1",
    "question": "What time is it?",
    "userId": "user1",
  },
  Object {
    "correlationId": "a123",
    "count": 0,
    "key": "2",
    "question": "What is for lunch?",
    "userId": "1",
  },
]
`;

exports[`fetchQuestions should fetch the questions for the given correlation id and set them in the state 2`] = `
Array [
  Array [
    "/questions?correlationId=a123",
  ],
]
`;

exports[`incrementQuestionCount should post the increment request to the backend 1`] = `
Array [
  Array [
    "/ask",
    Object {
      "body": "{\\"key\\":\\"question-key\\"}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`incrementQuestionCount should post the increment request to the backend with the expected metadata 1`] = `
Array [
  Array [
    "/ask",
    Object {
      "body": "{\\"key\\":\\"question-key\\",\\"userId\\":\\"my-user-id\\"}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;
